crates/proc_macro_api/src/lib.rs

```diff
@@ -108,4 +121,61 @@ impl ProcMacroClient {
             }
         }
     }
+
+    // This is used inside self.read_version() to locate the ".rustc" section
+    // from a proc macro crate's binary file.
+    fn read_section(&self, dylib_binary: &[u8], section_name: &str) -> &[u8] {
+        BinaryFile::parse(dylib_binary)
+            .unwrap()
+            .section_by_name(section_name)
+            .unwrap()
+            .data()
+            .unwrap()
+    }
+
+    // Check the version of rustc that was used to compile a proc macro crate's
+    // binary file.
+    // A proc macro crate binary's ".rustc" section has following byte layout:
+    // * [b'r',b'u',b's',b't',0,0,0,5] is the first 8 bytes
+    // * ff060000 734e6150 is followed, it's the snappy format magic bytes,
+    //   means bytes from here(including this sequence) are compressed in
+    //   snappy compression format. Version info is here inside, so decompress
+    //   this.
+    // The bytes you get after decompressing the snappy format portion has
+    // following layout:
+    // * [b'r',b'u',b's',b't',0,0,0,5] is the first 8 bytes(again)
+    // * [crate root bytes] next 4 bytes is to store crate root position,
+    //   according to rustc's source code comment
+    // * [length byte] next 1 byte tells us how many bytes we should read next
+    //   for the version string's utf8 bytes
+    // * [version string bytes encoded in utf8] <- GET THIS BOI
+    // * [some more bytes that we don really care but still there] :-)
+    // Check this issue for more about the bytes layout:
+    // https://github.com/rust-analyzer/rust-analyzer/issues/6174
+    fn read_version(&self, dylib_path: &Path) -> String {
+        let dylib_binary = fsread(dylib_path).unwrap();
```

Comment:
You should probably use `memmap` here and disable most of the `object` features (see https://github.com/rust-analyzer/rust-analyzer/pull/6817).

Reply:
I guess you mean memmap is more lightweight for task like just getting section bytes out?

Reply:
Yes, it maps the file, while you're reading the whole thing in memory. And if we disable the `unaligned` feature of `object`, a `Vec<u8>` might not always work (but I'm not sure).

Reply:
I tried to read the doc of `memmap` but I am not sure how to read a section of a file after mapping the file?

Reply:
Um, it's as easy as:

```rust
let file = File::open(file)?;
let buffer = unsafe { Mmap::map(&file)? };
let file = object::File::parse(&buffer);
```

(also taken from #6817)

Reply:
Haha. Looks like I totally misunderstood ur comment. :sweat_smile: 

So by this:

> You should probably use `memmap` here and disable most of the `object` features (see #6817).

what u actually saying is, instead of `std::fs::read`ing binary into a `Vec<u8>`, simply make a memory map of the file, then continue whatever section reading task. Yes?

(I thought u were saying don't use `object` crate at all, just use `memmap` and implement the whole binary parsing and seciton parsing logic from there. :sweat_smile: I was almost going to ask u what specific binary format are we supporting )

Reply:
Yup :-).

Reply:
About disabling features, I don't see anything on `object`'s doc page regarding what features to keep turned on for what functionalities I use. Should I do as u did in ur draft PR? (Also how would u figure out what features to on and off if there is no doc about that?)

Reply:
You can use the same ones if they work in your `object` version (I used a `git` dependency for a new feature).

I enabled `std` to get `object::Error` to implement `std::error:Error` (otherwise the code doesn't compile), `read_core` for reading anything (I assume) PE for Windows, ELF for Linux and Mach-O for Mac. I thought neither COFF support nor unaligned reading were required, but I just noticed that the Windows tests are failing, so maybe I still need one of those.